gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java

```diff
@@ -169,13 +169,35 @@ private String rangeSql(TimeSpan ts) {
     return format(RANGE_SQL, tableName("vals"), ts.start, ts.end);
   }
 
-  public ListenableFuture<Stats> getStats(TimeSpan span) {
-    return transform(expectOneRow(qe.query(statsSql(span))),
-        row -> new Stats(span, row.getDouble(0), row.getDouble(1), row.getDouble(2)));
+  private String statsSql(TimeSpan ts) {
+    return format(STATS_SQL, tableName("vals"), ts.start, ts.end);
   }
 
-  private String statsSql(TimeSpan span) {
-    return format(STATS_SQL, counter.id, span.start, span.end);
+  public ListenableFuture<Stats> getStats(TimeSpan span) {
```

Comment:
As the selected time span could be large and we are not doing any windowing here, please move the averaging to the SQL query, so we don't have to waste time transferring all that data.

Reply:
The averaging used to be done in the SQL query, that was the problem that was reported in the bug. We need to switch to a weighted average, so the standard SQL average is not suitable. Do you know how to implement a weighted average in SQL? I'm not enough of an SQL wizard to know how!

Reply:
I tested the performance of this on a standard 2 second trace and had no issue on a macbook with selections up to 99.x% of the full trace duration

Reply:
2 second traces are not exactly representative...

The following also has the benefit of working for counters with differing intervals between samples:
```
select min(value), max(value), sum(value * dur) / sum(dur) avg from (
  select min(dur, ts + dur - $start, $end + 1 - ts) dur, value from (
    select ts, lead(ts, 1, (select end_ts from trace_bounds)) over (order by ts) - ts dur, value
    from counter where track_id = $track)
  where  ts + dur >= $start and ts <= $end)
```

You can use the query viewer to test out your own queries.

Reply:
I looked at this again, and realized that we already have a SQL view that takes care of the inner-most part (and also distinguishes correctly for delta vs event type counters). We should use that view, so the query should be:
```
select min(value), max(value), sum(value * dur) / sum(dur) avg from
(select min(dur, ts + dur - $start, $end + 1 - ts) dur, value from $table where ts + dur >= $start and ts <= $end)
```
where `$table` is `tableName("vals")`.

Reply:
I just came here to comment I've tried the SQL you originally posted and it generates incorrect numbers. I'm not proficient enough in SQL to spot the bug though.

Is what you're proposing here just a refactor or does it correct the computation as well?

Looking at the numbers generated by the old SQL it appears to be doing a weighted average of the wrong data, or computing the weights wrong.

Reply:
They looked OK to me, when I was testing it. Can you give me a trace, counter, and specific time range, where you think the numbers are wrong? I can then try and fix it.